### DISCUSSION TITLE
add icon-color to Tile card

### DISCUSSION THREAD

**Discussion by Mariusthvdb** [Mariusthvdb/custom-icon-color#3] [Ideas]
continuing from https://github.com/Mariusthvdb/custom-icon-color/discussions/1#discussioncomment-8084524
seems the logical progression, with Tile card getting more and more popular to, have this plugin also colorize the icon on that card.

I copied the code used for the button card and edited, but this is not working yet ofc.
```
// Install a hook to update the Tile card with custom styling
  installTileCardStylingHook() {
    customElements.whenDefined("hui-tile-card").then(() => {
        const tileCard = customElements.get("hui-tile-card");
        if (!tileCard) return;
        if (tileCard.prototype?.updated) {
            const originalUpdated = tileCard.prototype.updated;
            tileCard.prototype.updated = function customUpdated(changedProps) {
                if (!changedProps.has('_stateObj')) {
                    return;
                }
                const { _stateObj } = this;
                if (_stateObj.attributes?.icon_color) {
                    this.style?.setProperty('--tile-icon-color', _stateObj.attributes.icon_color);
                }
                originalUpdated.call(this, changedProps);
            }
        }
    });
  },
```

card_modding the Tile card is less obvious than other cards, so maybe there is much more to is getting this plugin to touch the Tile card too

<img width="1036" alt="Scherm­afbeelding 2024-01-11 om 19 43 26" src="https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/fd8dc0eb-8950-47e7-b5c0-d8f016409952">

Hope it can be done without too much effort, or complexity.

**Comment by elchininet** [comment]
@Mariusthvdb,
Try with this:

```javascript
customElements.whenDefined("hui-tile-card").then(() => {
    const tileCard = customElements.get("hui-tile-card");
    if (!tileCard) return;
    if (tileCard.prototype?.updated) {
        const originalUpdated = tileCard.prototype.updated;
        tileCard.prototype.updated = function customUpdated(changedProps) {
            if (
                !changedProps.has('_config') ||
                !changedProps.has('hass')
            ) {
                return;
            }
            const { _config, hass } = this;
            const entityId = _config?.entity;
            const states = hass?.states;
            const iconColor = states?.[entityId]?.attributes?.icon_color;
            if (iconColor) {
                this.style?.setProperty('--icon-primary-color', _stateObj.attributes.icon_color);
            }
            originalUpdated.call(this, changedProps);
        }
    }
});
```

**Reply by Mariusthvdb** [reply]
somehow this didnt work at first, and threw some inexplicable errors about card-mod or mini-graph-card....
which I dont have on that view.

I changed 1 thing in the `if (iconColor)`: use `iconColor` in stead of `_stateObj.attributes.icon_color
`
```
// Install a hook to update the Tile card with custom styling
  installTileCardStylingHook() {
    customElements.whenDefined("hui-tile-card").then(() => {
        const tileCard = customElements.get("hui-tile-card");
        if (!tileCard) return;
        if (tileCard.prototype?.updated) {
            const originalUpdated = tileCard.prototype.updated;
            tileCard.prototype.updated = function customUpdated(changedProps) {
                if (
                    !changedProps.has('_config') ||
                    !changedProps.has('hass')
                ) {
                    return;
                }
                const { _config, hass } = this;
                const entityId = _config?.entity;
                const states = hass?.states;
                const iconColor = states?.[entityId]?.attributes?.icon_color;
                if (iconColor) {
                    this.style?.setProperty('--icon-primary-color', iconColor);
                }
                originalUpdated.call(this, changedProps);
            }
        }
    });
  },
```

and now it works! 

this is using custom-icon-color (so no template is set on the bottom card)

<img width="413" alt="Scherm­afbeelding 2024-01-12 om 23 10 30" src="https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/1662793b-bf94-4186-ae72-79e44ab9b314">

If I add it to the custom-ui-only, which does support templates, colors appear, and at first sight this seem great. However, they dont update after initial loading. The icon does change correctly and continuously:

<img width="1376" alt="Scherm­afbeelding 2024-01-12 om 23 17 20" src="https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/82b8405e-6330-4878-bc34-b1a7752b1990">

sorry for the bigger image here but check the 2 buttons on the left side for Intercom volume and Garage hydro temp signal strength (these are correct), and the 2 tile cards for the same entity.

navigating away from the view and returning changes the icon to the correct color, and again freezes. The icon itself does change.

**Reply by Mariusthvdb** [reply]
wait, I now see this also is the case on the entity card..... and sensor card

somehow the updating on those cards is not in sync with the entities in the entities cards, or the button card, which are perfect.

**Reply by elchininet** [reply]
>I changed 1 thing in the if (iconColor): use iconColor in stead of _stateObj.attributes.icon_color 

Yes, that‘s right. I am not testing the code, just coding looking at the properties, so I am a bit blind about what will be the result. But your change is correct.

>If I add it to the custom-ui-only, which does support templates, colors appear, and at first sight this seem great. However, they dont update after initial loading. The icon does change correctly and continuously:

That means that the `updated` method is not triggered when the entitities change or the `changedProps` doesn‘t contain the proper elements in certain situations. The best way to debug that would be placing `console.logs` and check what is triggered and when.

For example:

```javascript
customElements.whenDefined("hui-tile-card").then(() => {
        const tileCard = customElements.get("hui-tile-card");
        if (!tileCard) return;
        if (tileCard.prototype?.updated) {
            const originalUpdated = tileCard.prototype.updated;
            tileCard.prototype.updated = function customUpdated(changedProps) {

                // Here
                console.log('updated method called', changedProps);

                if (
                    !changedProps.has('_config') ||
                    !changedProps.has('hass')
                ) {
                    return;
                }
                const { _config, hass } = this;
                const entityId = _config?.entity;
                const states = hass?.states;
                const iconColor = states?.[entityId]?.attributes?.icon_color;

                // And here
                console.log('Does iconColor exist?', states?.[entityId]?.attributes?.icon_color);

                if (iconColor) {
                    this.style?.setProperty('--icon-primary-color', iconColor);
                }
                originalUpdated.call(this, changedProps);
            }
        }
    });
```

That way you can detect what is failing, if the `update` method is not called when the entity changes, if the properties are not inside the `changedProps` parameter or the `icon_color` is not there.

**Reply by Mariusthvdb** [reply]
ok will do, but it must be an update thing, because reloading the view shows the correct icon_color...

could it be it update for some card types, and not for others, even though the update technique is identical?

**Reply by elchininet** [reply]
>could it be it update for some card types, and not for others, even though the update technique is identical?

It could be, that is something that needs to be debugged. Maybe the `update` method is not triggered at all or maybe it fails in one of the internal checks.

**Reply by Mariusthvdb** [reply]
and adding those console logs causes continuous:

<img width="1352" alt="Scherm­afbeelding 2024-01-13 om 17 35 34" src="https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/c7d05050-d626-4da3-ae5e-843a9e107221">

so, it does show the update method being called on each and every state change in the system,...... bu I cant see the other log at all. Guess I'll take out the first.

Btw, this is proof using those templates in custom-ui as I do now is not good, it is truly constantly updating, and that must take its toll somewhere...

**Reply by Mariusthvdb** [reply]
<img width="279" alt="Scherm­afbeelding 2024-01-13 om 17 41 51" src="https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/438241a1-67a2-40e0-a6c8-5e06308e0500">

guess we're after those undefined.s....

**Reply by Mariusthvdb** [reply]
couldnt it have to do with the way custom-ui is initialized?:

```
  installCustomHooks() {
    window.customUI.installTemplateAttributesHook();
    window.customUI.installEntityCardStylingHook();
    window.customUI.installButtonCardStylingHook();
    window.customUI.installTileCardStylingHook();
    window.customUI.installStateBadgeStylingHook();
  },

  async init() {
// Check if initialization has already been done
    if (window.customUI.initDone) return;

// Wait for the hass.states to be populated
    const main = window.customUI.lightOrShadow(document, "home-assistant");
    await new Promise((resolve) => {
      const intervalId = setInterval(() => {
        if (main?.hass?.states) {
          clearInterval(intervalId);
          resolve();
        }
      }, 100);
    });

// Install the hooks and mark initialization as done
    window.customUI.installCustomHooks();
    window.customUI.initDone = true;

// Push custom-ui information to a global list
    window.CUSTOM_UI_LIST = window.CUSTOM_UI_LIST || [];
    window.CUSTOM_UI_LIST.push({
      name: Name,
      version: `${Version} ${Description}`,
      url: Url
    });
  },

```

I remember that was changed to make it lighter on the system some time ago when we had:

```
  installClassHooks() {
    if (window.customUI.classInitDone)
      return;

    window.customUI.classInitDone = true;
    window.customUI.installStatesHook();
    window.customUI.installStateBadge();
  },

  init() {
    if (window.customUI.initDone)
      return;

    window.customUI.installClassHooks();

    const main = window.customUI.lightOrShadow(document, "home-assistant");
    if (!main.hass?.states) {
      window.setTimeout(window.customUI.init, 1000);
      return;
    }
```

and to be complete, this is the template function, which was recently changed to use deep query by means of the stringify function:
```
// Install a hook to update the states with template attributes
  installTemplateAttributesHook() {
    customElements.whenDefined("home-assistant").then(() => {
      const homeAssistant = customElements.get("home-assistant");
      if (!homeAssistant?.prototype?._updateHass) return;
      const originalUpdate = homeAssistant.prototype._updateHass;
      homeAssistant.prototype._updateHass = function update(obj) {
        const { hass } = this;
        if (obj.states) {
          Object.keys(obj.states).forEach(async (key) => {
            const entity = obj.states[key];
            if (!entity.attributes.templates) {
              return;
            }
            const newEntity =
              await window.customUI.maybeApplyTemplateAttributes(
                hass,
                obj.states,
                entity
              );
            if (hass.states && JSON.stringify(entity) !== JSON.stringify(hass.states[key])) {
              obj.states[key] = newEntity;
            } else if (JSON.stringify(entity) !== JSON.stringify(newEntity)) {
              obj.states[key] = newEntity;
            }
          });
        }
        originalUpdate.call(this, obj);
      };
    });
  },
```

My main issue with it hs always been it updates all, when any state changes in HA, and not only the entities that are in view. your console.log proofs that to be happening, though I can not note any problem in the frontend. 
Cant be good though.

**Reply by elchininet** [reply]
>and adding those console logs causes continuous:

The method is called sending the value of the `changedProps` attribute, open that object and check that the `_config` property is inside (I am seeing that the `hass` is) and also check if the `config` object contains the `entity` property and it has a value.

>Btw, this is proof using those templates in custom-ui as I do now is not good, it is truly constantly updating, and that must take its toll somewhere...

If you use a `platform: template` you can change it to a `platform: time_pattern` and update the sensor every minute or the time that you think that is better.

>guess we're after those undefined.s....

So, then it continues? You mentioned that you cant see the other log at all. Every time that you get an update log you should have one with the value of the `icon_color` (remember that not all the cards get the icon color from the same place, just make sure that you are loging the proper variable and it is better to do the debug one by one, don‘t put it in all the elements or you will get too much noise)

**Reply by Mariusthvdb** [reply]
> If you use a platform: template you can change it to a platform: time_pattern and update the sensor every minute or the time that you think that is better.

no, I meant a template in customize, like:
```
  customize:

    sensor.garage_hygro_temp_signal_strength:
      templates:
        icon_color: >
          if (state == 'unknown') return 'gray';
          if (state >= -50) return 'darkgreen';
          if (state >= -60) return 'green';
          if (state >= -67) return 'lightgreen';
          if (state >= -70) return 'gold';
          if (state >= -80) return 'orange';
          return 'red';
        icon: >
          if (state >= -50) return 'mdi:wifi-strength-4';
          if (state >= -60) return 'mdi:wifi-strength-3';
          if (state >= -67) return 'mdi:wifi-strength-2';
          if (state >= -70) return 'mdi:wifi-strength-1';
          if (state >= -80) return 'mdi:wifi-strength-outline';
          return 'mdi:wifi-strength-alert-outline';
```

this changes on each state changes, and believe that is very frequent. Same eg on a power sensor.

> don‘t put it in all the elements or you will get too much noise)

I have it only in the Tile card now...

**Reply by elchininet** [reply]
Then I supsect that it is not related to the `customize` but to the `_updateHass` custom function. Let me try to check the whole code in the repo.

**Reply by elchininet** [reply]
Yes, it is related to the `_updateHass` custom function. Juts do this:

```javascript
customElements.whenDefined("home-assistant").then(() => {
  const homeAssistant = customElements.get("home-assistant");
  if (!homeAssistant?.prototype?._updateHass) return;
  const originalUpdate = homeAssistant.prototype._updateHass;
  homeAssistant.prototype._updateHass = function update(obj) {
    console.log('update');
    originalUpdate.call(this, obj);
  };
});
```

And check how many times `update` is triggered. It seems that Home Assistant calls this method on every update to any entity.

**Reply by Mariusthvdb** [reply]
where did you get that? I am using this currently:
```
// Install a hook to update the states with template attributes
  installTemplateAttributesHook() {
    customElements.whenDefined("home-assistant").then(() => {
      const homeAssistant = customElements.get("home-assistant");
      if (!homeAssistant?.prototype?._updateHass) return;
      const originalUpdate = homeAssistant.prototype._updateHass;
      homeAssistant.prototype._updateHass = function update(obj) {
        const { hass } = this;
        if (obj.states) {
          Object.keys(obj.states).forEach(async (key) => {
            const entity = obj.states[key];
            if (!entity.attributes.templates) {
              return;
            }
            const newEntity =
              await window.customUI.maybeApplyTemplateAttributes(
                hass,
                obj.states,
                entity
              );
            if (hass.states && JSON.stringify(entity) !== JSON.stringify(hass.states[key])) {
              obj.states[key] = newEntity;
            } else if (JSON.stringify(entity) !== JSON.stringify(newEntity)) {
              obj.states[key] = newEntity;
            }
          });
        }
        originalUpdate.call(this, obj);
      };
    });
  },
```
in custom-ui-only

but I added it there.......

https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/013116cd-2109-47d9-94dd-f34945d6cdbd

now you understand why I have been looking for a way to only update on entities that actually have a customization template set, and not on any state change in the system

**Reply by Mariusthvdb** [reply]
but anyways, this also indicates its not the Updating that is the issue, but the template not being applied to certain cards.

there is also a difference between reloading the view by clicking on the reload icon in the address bar (and only some of the cards get the icon color) or, navigating away and returning (when all of the icon_colors are correctly set)

**Reply by elchininet** [reply]
>but anyways, this also indicates its not the Updating that is the issue, but the template not being applied to certain cards.

No of course, that is not related to the issue that you had but you were interested in the recurrent number of times that method was executed.

To detect the issue you to check what I mentioned earlier:

>The method is called sending the value of the changedProps attribute, open that object and check that the _config property is inside (I am seeing that the hass is) and also check if the config object contains the entity property and it has a value.

**Reply by elchininet** [reply]
>now you understand why I have been looking for a way to only update on entities that actually have a customization template set, and not on any state change in the system

For this, you can use the same method that `kiosk-mode` uses: https://github.com/NemesisRE/kiosk-mode/blob/c4cde606ea6f8ccfe6ebb0ea67f9c26bc839535a/src/kiosk-mode.ts#L674-L695

**Reply by Mariusthvdb** [reply]
uhm, that is remotely understandable to me, and I do see what it tries to do.
However, way beyond my coding skills (and ive never used those protected functions.

```
event.event_type === STATE_CHANGED_EVENT &&
			entities.includes(event.data.entity_id)
```

looks exactly like what we're after!

I suppose you couldnt have a look in my current code of custom-ui-only and suggest a PR to do so?
it would be the most amazing update ever since the initial release many years ago.....

**Reply by elchininet** [reply]
Let me know in which repo you need it and I will try to simulate the same. But I need you to test if it works because you have the set up ready for that.

**Reply by Mariusthvdb** [reply]
it is for the custom-ui repo, and specifically for the current [custom-ui-only](https://github.com/Mariusthvdb/custom-ui/blob/master/custom-ui-only.js) (will be renamed soon to custom-ui, and the existing custom-ui will be phased out, as it still uses more-info functionality we now modernized in custom more-info)

this repo is the plugin that uses the customize templates, so needs the updating

**Comment by Mariusthvdb** [comment]
> To detect the issue you to check what I mentioned earlier:
> 
> The method is called sending the value of the changedProps attribute, open that object and check that the _config property is inside (I am seeing that the hass is) and also check if the config object contains the entity property and it has a value.

its impossible..... the list of those Updates is endless and super fast..

I took out the first console and changed the second to

`console.log('Does iconColor exist?', states?.[entityId], states?.[entityId]?.attributes?.icon_color);`

and that has some funny but interesting results, though I need to refresh the view for them to keep logging (is that a cue in itself;f?)

<img width="1443" alt="Scherm­afbeelding 2024-01-15 om 13 30 03" src="https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/2f95aa41-d41d-47e3-98bb-cb8a8b484345">

so the top listed shows an entity_id and an icon_color,
second shows green, but `...` and no entity_id,
and the 3d shows and entity, but is undefined....


I believe the entity's that are listed without a color are the ones not being used on this view, (and they dont have a customized template  either)

It would be good if we could stop them from being evaluated in the first place.

the listed colors without an entity are a mystery to me yet. its happening on the `input_number.intercom_volume` in the test view, but as the logging doesnt list it.


It does use the experimental jinja style of setting an icon_color (opposed to the customize javascript templates):
```
      - unique_id: intercom_volume
        state: >
          {{states('input_number.intercom_volume')}}
        icon: >
          {% set state = this.state|float(0) %}
          {% if state == 0.0 %} mdi:volume-off
          {% elif state <= 0.3 %} mdi:volume-low
          {% elif state <= 0.6 %} mdi:volume-medium
          {% else %} mdi:volume-high
          {% endif %}
# experimental
        attributes:
          icon_color: >
            {% set vol = this.state|float(0) %}
            {% if vol == 0.0 %} gray
            {% elif vol <= 0.3 %} dodgerblue
            {% elif vol <= 0.6 %} green
            {% elif vol <= 0.8 %} orange
            {% else %} red
            {% endif %}
```

Hoewever, I cant imagine that would be the cause. Because first of all, the other cards are not either by this (remember, entities card and button card are spot on), and secondly, some of the entities that do use a javascript icon_color template are also not displayed correctly on the tile card....

**Reply by elchininet** [reply]
Log it in this way:

```javascript
console.log('Does iconColor exist?', entityId, JSON.stringify(states?.[entityId]?.attributes));
```

And check the result.

**Reply by elchininet** [reply]
You can open a separate discussion for this to avoid mixing things.

>now you understand why I have been looking for a way to only update on entities that actually have a customization template set, and not on any state change in the system

I have checked your code. The update is done in entities that have a `template` property in their attributes, if the entity doesn‘t have that property [the code doesn‘t continue](https://github.com/Mariusthvdb/custom-ui/blob/5c708285c493526a0bd7bb423e37a97d85f3b32a/custom-ui-only.js#L70). So the reason about those entities being updated is not in that code.

**Reply by Mariusthvdb** [reply]
happening on a fresh loading of the view in the dashboard (icons get colored):

<img width="1379" alt="Scherm­afbeelding 2024-01-15 om 22 44 07" src="https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/14a153bc-df39-445a-8915-e8c70ae927c5">

<img width="538" alt="Scherm­afbeelding 2024-01-15 om 22 45 18" src="https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/6435251a-f15d-4fe3-89e6-caab8276446e">

this is on a reload of the same view:

<img width="540" alt="Scherm­afbeelding 2024-01-15 om 22 45 47" src="https://github.com/Mariusthvdb/custom-icon-color/assets/33354141/3cf9d633-e685-4bde-a889-f2b9926dd918">

and nothing logged.
I need to go out of the dashboard (left side menu) and come back again for the Tile cards to again update their color

**Reply by Mariusthvdb** [reply]
> I have checked your code. The update is done in entities that have a template property in their attributes, if the entity doesn‘t have that property [the code doesn‘t continue](https://github.com/Mariusthvdb/custom-ui/blob/5c708285c493526a0bd7bb423e37a97d85f3b32a/custom-ui-only.js#L70). So the reason about those entities being updated is not in that code.

Yes I have stared at that too... the thing that makes me wonder how and why though, is that it does actually work properly when used on entities in an entities card (StateBadge) and the Button card now. they are updated on each and every change that induces another color

**Reply by elchininet** [reply]
Hey @Mariusthvdb,

About this:

>happening on a fresh loading of the view in the dashboard (icons get colored)
>this is on a reload of the same view:
>and nothing logged.
>I need to go out of the dashboard (left side menu) and come back again for the Tile cards to again update their color

Could be the case that the code is not executed in those situations? Place a `console.log` in `installCustomHooks` and check if it is being called in all those cases.

**Reply by Mariusthvdb** [reply]
what should I add there?
```
// Install the hooks for updating states, entity cards, and state badges
  installCustomHooks() {
    window.customUI.installEntityCardStylingHook();
    window.customUI.installButtonCardStylingHook();
    window.customUI.installTileCardStylingHook();
    window.customUI.installStateBadgeStylingHook();
  },
```

**Reply by elchininet** [reply]
>what should I add there?

Just a `console.log` with some random text to check if it is called in all those cases.

**Comment by chillymattster** [comment]
Hi,
I'm not 100% sure if I understood everything of the ongoing discussion, but I after reading everything and spending some time, I found a way that enables automatic update if tile card. I created a pull request to share my work 😃 

From my understanding the problem is, that on every update of the ```hass```-object there is an additional check for an updated config as well. But the config is only updated once and then kept "as is" which prevents further updates of the icon color.
By changing the checks for an updated ```hass```-object and just the existence of a config, automatic updates are possible.

Furthermore, just setting ```--icon-primary-color``` changes the default behaviour of the tile card. Using ```--tile-color``` of ```ha-card```inside the tile card component leads to color changes that are visibly equal to the default behaviour.
